File: search_rotated.py
def search_rotated():
    nums = [4,5,6,7,0,1,2]
    target = 2

    l, r = 0, len(nums)-1
    while l <= r:
        mid = (l + r) // 2

        if nums[mid] == target:
            return mid

        if nums[mid] >= nums[l]:
            if nums[mid] < target or nums[l] > target:
                l = mid + 1
            else:
                r = mid - 1
        else:
            if nums[mid] > target or nums[r] < target:
                r = mid - 1
            else:
                l = mid + 1

    return -1

print(search_rotated())

File: search_rotated.rs

fn search_rotated(nums: Vec<i32>, target: i32) -> i32 {
    let (mut l, mut r) = (0, nums.len() - 1);

    while l <= r {
        let m = (l + r) / 2;

        if nums[m] == target {
            return m as i32;
        }

        if nums[m] >= nums[l] {
            if nums[m] < target || nums[l] > target {
                l = m + 1
            } else {
                r = m - 1
            }
        }
        else {
            if nums[m] > target || nums[r] < target {
                r = m - 1
            } else {
                l = m + 1
            }
        }
    }
    -1
}

fn main() {
    let nums: Vec<i32> = vec![4,5,6,7,0,1,2];
    let target = 2;
    println!("{:?}", search_rotated(nums, target));
}
